src/aead/chacha20.rs

```diff
@@ -0,0 +1,155 @@
+
+        // construct a TLSInnerPlaintext
+        let mut payload = Vec::with_capacity(total_len);
+        payload.extend_from_slice(m.payload);
+        payload.push(m.typ.get_u8());
+
+        let nonce = chacha20poly1305::Nonce::from(cipher::Nonce::new(&self.1, seq).0);
+        let aad = cipher::make_tls13_aad(total_len);
+
+        self.0
+            .encrypt_in_place(&nonce, &aad, &mut payload)
+            .map_err(|_| rustls::Error::EncryptError)
+            .map(|()| {
+                cipher::OpaqueMessage::new(
+                    ContentType::ApplicationData,
+                    ProtocolVersion::TLSv1_2,
+                    payload,
+                )
+            })
+    }
+}
+
+impl cipher::MessageDecrypter for Tls13Cipher {
+    fn decrypt(
+        &self,
+        mut m: cipher::OpaqueMessage,
+        seq: u64,
+    ) -> Result<cipher::PlainMessage, rustls::Error> {
+        let payload = m.payload_mut();
+        let nonce = chacha20poly1305::Nonce::from(cipher::Nonce::new(&self.1, seq).0);
+        let aad = cipher::make_tls13_aad(payload.len());
+
+        self.0
+            .decrypt_in_place(&nonce, &aad, payload)
+            .map_err(|_| rustls::Error::DecryptError)?;
+
+        m.into_tls13_unpadded_message()
+    }
+}
+
+#[cfg(feature = "tls12")]
+struct Tls12Cipher(chacha20poly1305::ChaCha20Poly1305, cipher::Iv);
+
+#[cfg(feature = "tls12")]
+impl cipher::MessageEncrypter for Tls12Cipher {
+    fn encrypt(
+        &self,
+        m: cipher::BorrowedPlainMessage,
+        seq: u64,
+    ) -> Result<cipher::OpaqueMessage, rustls::Error> {
+        let total_len = m.payload.len() + CHACHAPOLY1305_OVERHEAD;
+
+        let mut payload = Vec::with_capacity(total_len);
+        payload.extend_from_slice(m.payload);
+
+        let nonce = chacha20poly1305::Nonce::from(cipher::Nonce::new(&self.1, seq).0);
+        let aad = cipher::make_tls12_aad(seq, m.typ, m.version, payload.len());
+
+        self.0
+            .encrypt_in_place(&nonce, &aad, &mut payload)
+            .map_err(|_| rustls::Error::EncryptError)
+            .map(|_| cipher::OpaqueMessage::new(m.typ, m.version, payload))
+    }
+}
+
+#[cfg(feature = "tls12")]
+impl cipher::MessageDecrypter for Tls12Cipher {
+    fn decrypt(
+        &self,
+        mut m: cipher::OpaqueMessage,
+        seq: u64,
+    ) -> Result<cipher::PlainMessage, rustls::Error> {
+        let payload = m.payload();
+        let nonce = chacha20poly1305::Nonce::from(cipher::Nonce::new(&self.1, seq).0);
+        let aad = cipher::make_tls12_aad(
+            seq,
+            m.typ,
+            m.version,
+            payload.len() - CHACHAPOLY1305_OVERHEAD,
+        );
+
+        let payload = m.payload_mut();
+        self.0
+            .decrypt_in_place(&nonce, &aad, payload)
+            .map_err(|_| rustls::Error::DecryptError)?;
+
+        Ok(m.into_plain_message())
+    }
+}
+
+const CHACHAPOLY1305_OVERHEAD: usize = 16;
```

Comment:
This is available as [`chacha20poly1305::ChaCha20Poly1305::TagSize::USIZE`](https://docs.rs/chacha20poly1305/latest/chacha20poly1305/struct.ChaChaPoly1305.html#associatedtype.TagSize) FWIW, though it would require the [`AeadCore`](https://docs.rs/aead/latest/aead/trait.AeadCore.html) and [`Unsigned`](https://docs.rs/typenum/latest/typenum/marker_traits/trait.Unsigned.html) traits in scope.

Reply:
```
use chacha20poly1305::{AeadInPlace, KeyInit, KeySizeUser, ChaCha20Poly1305::TagSize::USIZE as CHACHAPOLY1305_OVERHEAD};

failed to resolve: `ChaCha20Poly1305` is a type alias, not a module
`ChaCha20Poly1305` is a type alias, not a module
```
Strange

Reply:
Actually, I think I did use that before which is exactly this problem.

Reply:
It may require `<chacha20poly1305::ChaCha20Poly1305 as AeadCore>::TagSize`. I can take a look.

Reply:
Yeah, that works IIRC from before, but I think this is pretty verbose, while the TLS is not making this variable at any point, I would say hardcoding it to 16 would be fine...